### PR TITLE
clang for travis

### PR DESCRIPTION
**Pull request by argakiig**
build clang and gcc based ci docker containers
split deployment out to a separate stage

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,50 +1,56 @@
...
-  - if [ "$TRAVIS_OS_NAME" = "linux" ]; then docker run -v $TRAVIS_BUILD_DIR:/workspace -v $HOME/.ccache:/ccache -v $HOME/.cargo:/cargo nanocurrency/nano-ci /bin/bash -c "apt install ccache; cd /workspace && ASAN=${ASAN} TSAN=${TSAN} CCACHE_DIR=/ccache CARGO_HOME=/cargo ./ci/build-travis.sh /usr/lib/x86_64-linux-gnu/cmake/Qt5 ${PWD}"; fi
-
-deploy:
-  provider: script
-  skip_cleanup: true
-  script: ci/deploy-docker.sh
-  on:
-    all_branches: true
-    condition: -n "$ONE_TIME_TESTS"
+  - if [ "$TRAVIS_OS_NAME" = "linux" ]; then docker run -v $TRAVIS_BUILD_DIR:/workspace -v $HOME/.ccache:/ccache -v $HOME/.cargo:/cargo nanocurrency/nano-ci-$TRAVIS_COMPILER /bin/bash -c "apt install ccache; cd /workspace && ASAN=${ASAN} TSAN=${TSAN} CCACHE_DIR=/ccache CARGO_HOME=/cargo ./ci/build-travis.sh /usr/lib/x86_64-linux-gnu/cmake/Qt5 ${PWD}"; fi
```
Comment (wezrule) [inline .travis.yml]: Should probably have a new line at the end of the file

File: docker/ci/Dockerfile-clang [hunk 2]
```diff
@@ -0,0 +1,37 @@
+FROM ubuntu:16.04
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update -qq && apt-get install -yqq \
+    build-essential \
+    cmake \
+    g++ \
```
Comment (wezrule) [inline docker/ci/Dockerfile-clang]: is g++ required for a clang build?

File: docker/ci/Dockerfile-clang [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+RUN apt-get update -qq && apt-get install -yqq \
+    clang-7 lldb-7 lld-7 libc++-7-dev libc++abi-7-dev 
+
+RUN update-alternatives --install /usr/bin/cc cc /usr/bin/clang-7 100
+RUN update-alternatives --install /usr/bin/c++ c++ /usr/bin/clang++-7 100
+RUN ln -s /usr/bin/clang-7 /usr/bin/clang
+RUN ln -s /usr/bin/clang++-7 /usr/bin/clang++
+ENV CXX=clang++
+ENV CC=clang
+RUN ./bootstrap_boost.sh -m -c -B 1.69
```
Comment (wezrule) [inline docker/ci/Dockerfile-clang]: We should probably use the minimum supported boost version (default 1.69)

File: docker/ci/Dockerfile-clang [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    software-properties-common
+
+RUN wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+
+run apt-add-repository "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main"
+
+ENV BOOST_ROOT=/usr/local
+ADD util/build_prep/bootstrap_boost.sh bootstrap_boost.sh
+RUN apt-get update -qq && apt-get install -yqq \
+    clang-7 lldb-7 lld-7 libc++-7-dev libc++abi-7-dev
```
Comment (wezrule) [inline docker/ci/Dockerfile-clang]: should we be using such a new version of clang? A minimum requirement version may show more errors.

File: .travis.yml (hunk 1 above)
Comment (argakiig) [reply .travis.yml]: I always lose them, added, thanks

File: docker/ci/Dockerfile-clang (hunk 2 above)
Comment (argakiig) [reply docker/ci/Dockerfile-clang]: This was left over from the previous docker image, requirements have been adjusted for clang specifically

File: docker/ci/Dockerfile-clang (hunk 3 above)
Comment (argakiig) [reply docker/ci/Dockerfile-clang]: removed explicit 1.69 fell back to 1.67

File: docker/ci/Dockerfile-clang (hunk 4 above)
Comment (argakiig) [reply docker/ci/Dockerfile-clang]: initially tried clang3.8 which is pulled down with clang, discovered needed 3.9 updated to use 3.9 and symlink that instead

**Pull request merged**